File: epi_judge_cpp/search_in_list.cc
#include <memory>

#include "list_node.h"
#include "test_framework/generic_test.h"
using std::shared_ptr;

shared_ptr<ListNode<int>> SearchList(shared_ptr<ListNode<int>> L, int key) {
  // TODO - you fill in here.
  return nullptr;
}
int SearchListWrapper(shared_ptr<ListNode<int>> L, int key) {
  auto result = SearchList(L, key);
  return result ? result->data : -1;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L", "key"};
  return GenericTestMain(args, "search_in_list.cc", "search_in_list.tsv",
                         &SearchListWrapper, DefaultComparator{}, param_names);
}

File: epi_judge_java_solutions/epi/SearchInList.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class SearchInList {

  public static ListNode<Integer> searchList(ListNode<Integer> L, int key) {

    while (L != null && L.data != key) {
      L = L.next;
    }
    // If key was not present in the list, L will have become null.
    return L;
  }

  @EpiTest(testDataFile = "search_in_list.tsv")
  public static int searchListWrapper(ListNode<Integer> L, int key) {
    ListNode<Integer> result = searchList(L, key);
    return result != null ? result.data : -1;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SearchInList.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
